server/src/main/java/org/opensearch/cluster/service/MasterTaskThrottler.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.service;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.Version;
+import org.opensearch.action.admin.cluster.snapshots.delete.DeleteSnapshotRequest;
+import org.opensearch.action.admin.indices.create.CreateIndexRequest;
+import org.opensearch.action.admin.indices.dangling.delete.DeleteDanglingIndexRequest;
+import org.opensearch.action.admin.indices.mapping.put.PutMappingClusterStateUpdateRequest;
+import org.opensearch.common.metrics.CounterMetric;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.snapshots.UpdateIndexShardSnapshotStatusRequest;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This class is extension of {@link Throttler} and does throttling of master tasks.
+ *
+ * This class does throttling on task submission to master, it uses class name of request of tasks as key for
+ * throttling. Throttling will be performed over task class level, different task types have different request class, i.e.
+ * Throttling is at the task type level.
+ *
+ * Set "master.throttling.enable" setting to enable/disable this throttling. (Default False)
+ * Set specific setting to for setting the threshold of throttling of particular task type.
+ * e.g : Set "master.throttling.thresholds.put_mapping" to set throttling limit of "put mapping" tasks,
+ *       Set it to default value(-1) to disable the throttling for this task type.
+ *
+ *       We can also provide the class path of request, which we want to throttle. If new task type has introduced and we
+ *       haven't added that in configuration then we can use the class path of request to enable throttling on that type.
+ *       e.g use "org/opensearch/action/admin/indices/create/CreateIndexClusterStateUpdateRequest" as key in setting
+ *       throttling limit for Create index tasks.
```

Comment:
[Old PR Comment](https://github.com/opensearch-project/OpenSearch/pull/484#discussion_r611466176)

How will this setting key look like in cluster state? What happens in case as part of some refactoring we move around classes? Then, all of a sudden these settings would start failing? or are we going to migrate these settings automatically?

Also it feels like a backdoor, what if user keeps on provide non throttling applicable classes here, how do you control that?

Reply:
>How will this setting key look like in cluster state?

Key will looks same in cluster setting. Like "org/opensearch/action/admin/indices/create/CreateIndexClusterStateUpdateRequest".

>What happens in case as part of some refactoring we move around classes? Then, all of a sudden these settings would start failing?

Yes if class are refactored it would start failing. Ideally new node will not come up as prepareCommit it self fails to apply setting. 
For refactoring cases, we will need to remove old settings, move cluster to new release with refactored class and then reapply with new class path. 
Do you have any suggestion to improve this behavior? Will be happy to improve if there is any scope. 

>Also it feels like a backdoor, what if user keeps on provide non throttling applicable classes here, how do you control that?

Nice Suggestion! 
I didn't had any verification for it, but I have added it now. While applying setting we will check class should be MasterNode Request or ClusterState Update Request. Through this validation we can control it.  Please let me know If I have missed any other request type in check. Thanks.

Reply:
@dhwanilpatel : I am thinking why do we even need this? Anytime a new class is added, it should be caught in the review and a decision should be taken whether it should be added in throttler or not.

Reply:
Your suggestion is good, I think it will be cleaner way as well. Operator don't have to worry about class path. I will evaluate all the pending tasks and add those here in this mapping where we may need throttling.

Just one concern, Is there any way to make sure we caught new tasks in PR and evaluate requirement of throttling?

Reply:
I have found one for issue while doing above change, for most of the task submission it is using anonymous class, so can't rely on class name of submitted task for those.  

I have changed logic little bit now. Now we will do throttling based on executor type. Each task has its own executor implementation so relationship would be 1:1 only.  I have added one method to get the throttling key in `ClusterStateTaskExecutor`. For tasks where we want to configure throttling they will override it and provide the key for throttling. 

I have analyzed the various task types and added key for most of the tasks which are generated from customer like create index/put mapping/etc. Please validate that if I have covered enough tasks or not.